### ISSUE TITLE
Acronyms with periods at the end of the sentence are tokenized incorrectly

### ISSUE THREAD

**Issue by BLKSerene** [nltk/nltk#3035]
If acronyms with periods are at the end of the sentence, `TreebankWordTokenizer` and `NLTKWordTokenizer` would split the last period (which serves both as part of the acronym and as a full stop):

```
>>> import nltk
>>> nltk.TreebankWordTokenizer().tokenize('I have been to U.S.A.')
['I', 'have', 'been', 'to', 'U.S.A', '.']
>>> nltk.NLTKWordTokenizer().tokenize('I have been to U.S.A.')
['I', 'have', 'been', 'to', 'U.S.A', '.']
```

Sentences with acronyms at the start or middle are handled correctly:

```
>>> nltk.TreebankWordTokenizer().tokenize('U.S.A. is a north American country.')
['U.S.A.', 'is', 'a', 'north', 'American', 'country', '.']
>>> nltk.NLTKWordTokenizer().tokenize('U.S.A. is a north American country.')
['U.S.A.', 'is', 'a', 'north', 'American', 'country', '.']
```

**Comment by arademaker** [comment]
https://stackoverflow.com/questions/47274540/how-to-improve-nltk-sentence-segmentation

> The awesome-ness of Kiss and Strunk (2006) Punkt algorithm is that it's unsupervised. So given a new text, you should retrain the model and apply the model to your text

**Comment by arademaker** [comment]
I don't think this is a code issue, so I would believe we can close this issue.

**Comment by BLKSerene** [comment]
@arademaker The problem concerns with the word tokenizers, not the Punkt sentence tokenizer. Are they based on the same algorithms?

**Comment by BLKSerene** [comment]
In my use cases, I do not use `nltk.word_tokenize` (which would call `nltk.sent_tokenize` first). I call `TreebankWordTokenizer` and `NLTKWordTokenizer` directly for word tokenization task (sentence tokenization, if needed, is handled separately by other functions).

**Comment by tomaarsen** [comment]
@arademaker this indeed does not concern Punkt. `NLTKWordTokenizer` and `TreebankWordTokenizer`, which are regex tokenizers. To be specific, the following regex causes the issue:
https://github.com/nltk/nltk/blob/3b72fa2b3dea0088cd6d15d4409ce52318a3d1a0/nltk/tokenize/treebank.py#L65-L68

It searches for occurrences of a non-dot, followed by a dot, followed by some token that may be after a dot (e.g. `"` or `)`), and then whitespace, before the end of the text. As a result, it matches `A.` in `I have been to U.S.A.`. 

This can be "fixed" (or rather, improved), by changing the regex such that the first non-dot is now a non-dot *and* non-capitalised letter. I.e. we get `([^\.A-Z])(\.)([\]\)}>"\']*)\s*$` as the regex. 
However, this is not ideal:
### Current situation
```python
>>> import nltk
>>> nltk.TreebankWordTokenizer().tokenize('I have been to U.S.A.')
['I', 'have', 'been', 'to', 'U.S.A', '.']
>>> nltk.TreebankWordTokenizer().tokenize('I have been to USA.')
['I', 'have', 'been', 'to', 'USA', '.']
```

### Situation after described fix
```python
>>> import nltk
>>> nltk.TreebankWordTokenizer().tokenize('I have been to U.S.A.')
['I', 'have', 'been', 'to', 'U.S.A.']
>>> nltk.TreebankWordTokenizer().tokenize('I have been to USA.')
['I', 'have', 'been', 'to', 'USA.']
```

In short, this would likely be a regression for most cases. 

Perhaps some more complex solutions could work, e.g. checking for texts that (don't) match `([A-Z]\.){2,}`. I'm open for someone to have a look at this, but be wary of the potential regressions that could follow.

**Comment by tomaarsen** [comment]
I'm wary that all solutions would introduce regressions, so I'll close this for now.
